Add format options for ansible-galaxy collection list

shanemcd opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request


##### COMPONENT NAME
`ansible-galaxy`

##### ADDITIONAL INFORMATION
This would be useful for an upcoming feature in AWX/Tower.

shanemcd committed:
Add format options for ansible-galaxy collection list

shanemcd committed:
Include all collections in single json object / yaml document

shanemcd committed:
Fix tests

jborean93 commented on lib/ansible/cli/galaxy.py:
```diff
@@ -301,6 +302,10 @@ def add_list_options(self, parser, parents=None):
         list_parser.add_argument(galaxy_type, help=galaxy_type.capitalize(), nargs='?', metavar=galaxy_type)
 
+        if galaxy_type == 'collection':
+            list_parser.add_argument('-o', '--output', dest='output_format',
```
might be a good idea to have a 3rd option like `default` or `human` which is the current behaviour. You can also have a `choices=[]` kwarg where you define the valid choices. This helps both on validation and people using argcomplete to tab complete the options.

jborean93 reviewed (changes requested):
I'm fine with the idea, we just need a changelog fragment and integration test for this. You can find the current tests at https://github.com/ansible/ansible/blob/devel/test/integration/targets/ansible-galaxy-collection/tasks/list.yml.

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1313,7 +1318,8 @@ def execute_list(self):
             self.execute_list_role()
         elif context.CLIARGS['type'] == 'collection':
-            self.execute_list_collection()
+            output_format = context.CLIARGS.get('output_format', False)
```
This should be moved inside `execute_list_collection()` for consistency — all other code accesses `context.CLIARGS` directly so it's unnecessary to have an injection here.

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1386,6 +1392,9 @@ def execute_list_collection(self, artifacts_manager=None):
         warnings = []
         path_found = False
         collection_found = False
+        if output_format is not None:
+            collections_in_paths = {}
```
Declaring a variable in only one of the code branches is dangerous because it's hard to maintain accessing it under the same conditions in other places which can become a real bug during refactoring.
```suggestion
        collections_in_paths = {}
```

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1451,13 +1460,18 @@ def execute_list_collection(self, artifacts_manager=None):
-                # Display header
-                fqcn_width, version_width = _get_collection_widths(collections)
-                _display_header(collection_path, 'Collection', 'Version', fqcn_width, version_width)
+                if output_format is not None:
```
I think that the first branch of `if collection_name` should also emit JSON. Also, with Jordan's suggestion to use the `choices` kwarg in the CLI declaration, it'd be something like `output_format == 'json'`.

shanemcd replied:
Not sure I'm following. We loop over `collections_search_paths`. We need to include all of these in the json that's emitted, so we can only do that once we have the entirety of the data that needs to be printed.

webknjaz replied:
Just think about `ansible-galaxy collection list ns.name -o yaml`. Currently, your logic skips this case and falls back to the human-readable format.

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1466,6 +1480,14 @@ def execute_list_collection(self, artifacts_manager=None):
         for w in warnings:
             display.warning(w)
 
+        if output_format is not None:
+            if output_format == 'json':
+                display.display(json.dumps(collections_in_paths))
+            elif output_format == 'yaml':
+                display.display(yaml.safe_dump(collections_in_paths))
+            else:
+                raise AnsibleError("Only yaml or json are supported output types")
```
This could be less nested once you apply `choices`.

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -12,6 +12,7 @@
 import textwrap
 import time
 import yaml
+import json
```
This is injected right between two `yaml` imports which is also a third-party. Please move it up, before line 8 (all other imports are sorted alphabetically).

webknjaz reviewed (changes requested):
This needs integration tests + some structural improvements:

shanemcd committed:
Address PR feedback

shanemcd commented:
@jborean93 @webknjaz I have addressed the majority of your feedback. However, I have been unsuccessful at getting the integration tests to work on my machine, so I cannot add a new one.

Using the command:

```
bin/ansible-test integration ansible-galaxy-collection -v
```

I see the error:

```
TASK [ansible-galaxy-collection : install simple collection with implicit path - galaxy_ng] **************************************************************************************************************
fatal: [testhost]: FAILED! => {"changed": true, "cmd": ["ansible-galaxy", "collection", "install", "namespace1.name1", "-s", "galaxy_ng", "-v"], "delta": "0:00:01.617742", "end": "2021-02-04 10:04:31.690424", "msg": "non-zero return code", "rc": 1, "start": "2021-02-04 10:04:30.072682", "stderr": "\u001b[0;31mERROR! Failed to download collection tar from 'galaxy_ng': <urlopen error [Errno -2] Name or service not known>\u001b[0m", "stderr_lines": ["\u001b[0;31mERROR! Failed to download collection tar from 'galaxy_ng': <urlopen error [Errno -2] Name or service not known>\u001b[0m"], "stdout": "\u001b[0;34mUsing /tmp/ansible.xmzexh7v.test/galaxy/ansible.cfg as config file\u001b[0m\nStarting galaxy collection install process\nProcess install dependency map\nStarting collection install process\nDownloading http://ansible-ci-pulp:80/pulp/content/published/namespace1-name1-1.0.9.tar.gz to /home/shanemcd/.ansible/tmp/ansible-local-139077gi10ktw5/tmpr1mb62q0/namespace1-name1-1.0.9-aer09auq", "stdout_lines": ["\u001b[0;34mUsing /tmp/ansible.xmzexh7v.test/galaxy/ansible.cfg as config file\u001b[0m", "Starting galaxy collection install process", "Process install dependency map", "Starting collection install process", "Downloading http://ansible-ci-pulp:80/pulp/content/published/namespace1-name1-1.0.9.tar.gz to /home/shanemcd/.ansible/tmp/ansible-local-139077gi10ktw5/tmpr1mb62q0/namespace1-name1-1.0.9-aer09auq"]}
```

I spent about an hour trying to figure this out and had 0 luck.

s-hertel commented:
These integration tests need to be run with --docker

shanemcd committed:
Add tests for new galaxy list output format options

shanemcd committed:
Add changelog fragment

webknjaz commented on test/integration/targets/ansible-galaxy-collection/tasks/list.yml:
```diff
@@ -31,6 +31,28 @@
+
+- assert:
+    that:
+      - "{{ item.value | length == 3 }}"
```
You don't need to render this through "mustache"
```suggestion
    - item.value | length == 3
```

webknjaz commented on test/integration/targets/ansible-galaxy-collection/tasks/list.yml:
```diff
@@ -31,6 +31,28 @@
+
+- assert:
+    that:
+      - "{{ item.value | length == 3 }}"
```
I'd expect the test to check the contents, not just length.

webknjaz commented on test/integration/targets/ansible-galaxy-collection/tasks/list.yml:
```diff
@@ -31,6 +31,28 @@
       - "'dev.collection3 *' in list_result.stdout"
 
+- name: list collections in yaml format
+  command: ansible-galaxy collection list -o yaml
```
I'd expect extra tests for listing specific collections too. Like `ansible-galaxy collection list -o yaml ns1.coll2` and `ansible-galaxy collection list -o yaml ns1.coll2:1.0` maybe...

shanemcd replied:
Added below

shanemcd committed:
Add more tests for collection list output feature

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1379,9 +1384,11 @@ def execute_list_collection(self, artifacts_manager=None):
         :param artifacts_manager: Artifacts manager.
         """
 
+        output_format = context.CLIARGS.get('output_format', False)
```
Since `default` + `choices` are used, it's guaranteed that the key exists and so there's no point in defaulting to `False`. There's no way it's useful. So let's not assume a missing key. If somebody refactors things and it'll start failing, then this must start failing loudly too.
```suggestion
        output_format = context.CLIARGS['output_format']
```

jborean93 replied:
Agree on this it should change.

shanemcd replied:
Done.

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1451,13 +1458,18 @@ def execute_list_collection(self, artifacts_manager=None):
-                # Display header
-                fqcn_width, version_width = _get_collection_widths(collections)
-                _display_header(collection_path, 'Collection', 'Version', fqcn_width, version_width)
+                if output_format in ('yaml', 'json'):
```
Semantically, it's better to use a set here:
```suggestion
                if output_format in {'yaml', 'json'}:
```

shanemcd replied:
Done.

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1451,13 +1458,18 @@ def execute_list_collection(self, artifacts_manager=None):
-                fqcn_width, version_width = _get_collection_widths(collections)
-                _display_header(collection_path, 'Collection', 'Version', fqcn_width, version_width)
+                if output_format in ('yaml', 'json'):
+                    collections_in_paths[collection_path] = [
```
Why add another level of nesting with a list? Just have a dict comprehension.

jborean93 replied:
In case we ever add more field to this like author name or something of the sort. Keeping it as a dict entry in a list gives us more flexibility in the future.

shanemcd replied:
After further consideration, I actually agree with this point. It makes looping over this data easier too, as seen in the tests. If we wanted to add more information to the sub-dictionary in the future, that's possible too. Thanks for this feedback.

webknjaz replied:
Well, if we want flexibility, we could have a dict value but I still expect it to we inside a dict, not inside a list.
It's `{'ns.coll': {'fqcn': 'ns.coll', 'ver': '1.2.0'}}` vs `[{'fqcn': 'ns.coll', 'ver': '1.2.0'}]`: with a list, it's harder to access a specific collection as you'd have to do a full list scan to find in which element your target is while with a dict you can access it by key instantly.

cc @shanemcd

webknjaz replied:
Oh, I see it's addressed now.

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1451,13 +1458,18 @@ def execute_list_collection(self, artifacts_manager=None):
+                else:
+                    # Display header
+                    fqcn_width, version_width = _get_collection_widths(collections)
+                    _display_header(collection_path, 'Collection', 'Version', fqcn_width, version_width)
 
-                # Sort collections by the namespace and name
-                for collection in sorted(collections, key=to_text):
-                    _display_collection(collection, fqcn_width, version_width)
+                    # Sort collections by the namespace and name
+                    for collection in sorted(collections, key=to_text):
+                        _display_collection(collection, fqcn_width, version_width)
```
Please dedent this back to keep the nesting sane. It's better to use `continue` to skip it when necessary instead.
```suggestion
                    continue

                # Display header
                fqcn_width, version_width = _get_collection_widths(collections)
                _display_header(collection_path, 'Collection', 'Version', fqcn_width, version_width)

                # Sort collections by the namespace and name
                for collection in sorted(collections, key=to_text):
                    _display_collection(collection, fqcn_width, version_width)
```

shanemcd replied:
Done.

webknjaz reviewed (changes requested):
Looking at the source, I'm pretty sure that `ansible-galaxy collection list ns.name -o json` will output a human-readable table followed by an empty JSON object `{}`. Please add a test for this and check that only JSON is returned.

jborean93 reviewed (approved):
Apart from some of the comments already made by @webknjaz I'm fairly happy with this PR.

shanemcd committed:
For galaxy list yaml/json output, use dictionary of dictionaries instead of list

shanemcd committed:
Add tests for listing single collection in yaml / output format

shanemcd committed:
Address remaining PR feedback

shanemcd commented:
> Looking at the source, I'm pretty sure that `ansible-galaxy collection list ns.name -o json` will output a human-readable table followed by an empty JSON object `{}`. Please add a test for this and check that only JSON is returned.

Not sure how I overlooked that branch in the code. I've fixed this and added some tests. Thanks.

webknjaz commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1430,6 +1437,13 @@ def execute_list_collection(self, artifacts_manager=None):
+                        collection.fqcn: { 'version': collection.ver }
+                    }
+
+                    continue
```
Please move this above the `_get_collection_widths()` call to avoid running that code unnecessarily. It'd be basically the same as what's done in the else-branch that has a similar code below this snippet.

shanemcd replied:
Done

webknjaz commented on test/integration/targets/ansible-galaxy-collection/tasks/list.yml:
```diff
@@ -31,6 +31,58 @@
+  with_dict: "{{ list_result_json.stdout | from_json }}"
+
+- name: list single collection in json format
+  command: "ansible-galaxy collection list {{ item.key }} -o json"
+  register: list_single_result_json
+  environment:
+    ANSIBLE_COLLECTIONS_PATH: "{{ galaxy_dir }}/dev:{{ galaxy_dir }}/prod"
+  with_dict: "{{ { 'dev.collection1': '*', 'dev.collection2': 'placeholder', 'dev.collection3': '*' } }}"
```
This is rather overengineered:
```suggestion
  command: "ansible-galaxy collection list {{ item }} -o json"
  register: list_single_result_json
  environment:
    ANSIBLE_COLLECTIONS_PATH: "{{ galaxy_dir }}/dev:{{ galaxy_dir }}/prod"
  loop:
  - dev.collection1
  - dev.collection2
  - dev.collection3
```

shanemcd replied:
Sorry if it seems that way. But if you look at the assertions below, I am using the value in this dictionary to access the expected version of the collection. If you have a suggestion on how to better restructure this test, I am happy to update. But this suggestion breaks the tests.

webknjaz replied:
@shanemcd I'm pretty it shouldn't break tests in this place. You don't use the values. The assertion below uses the results that come from the stdout output. In this specific task, you only use keys.

shanemcd replied:
@webknjaz I am using the values in the assertion directly proceeding this task. Notice ` == item.item.value`

webknjaz replied:
Yes, but that refers to what you feed to the assertion task's `with_dict`, nothing to do with this task.

webknjaz replied:
Oh... I see what you did. You use a closure. This is clever but it is also hard to follow.

webknjaz reviewed (changes requested):
Almost LGTM: I have a few final comments below.

Also, could you please add tests with explicit `ansible-galaxy collection list -o human`? I believe I've only seen an implicit one.

s-hertel reviewed (approved):
Nice addition!

shanemcd committed:
Fix linter failures

bcoca commented on lib/ansible/cli/galaxy.py:
```diff
@@ -301,6 +302,10 @@ def add_list_options(self, parser, parents=None):
         list_parser.add_argument(galaxy_type, help=galaxy_type.capitalize(), nargs='?', metavar=galaxy_type)
 
+        if galaxy_type == 'collection':
+            list_parser.add_argument('-o', '--output', dest='output_format', choices=('human', 'yaml', 'json'), default='human',
```
we use -o for output file elsewhere, it is not good to overload options with diff meanings

webknjaz replied:
Fair enough. How about `-f` / `--output-format`?

s-hertel replied:
Could use `--format` to be similar to pip. I wouldn't add `-f` though since it's already associated with `--force`.

webknjaz replied:
Right, missed that. How about `--output-as`?

shanemcd replied:
I think `--format` makes sense. Let's go with that. Thanks all.

shanemcd committed:
Fix tests

shanemcd committed:
PR feedback

shanemcd committed:
--output -> --format

shanemcd committed:
Add explicit test for listing collection in human format

shanemcd commented:
> Almost LGTM: I have a few final comments below.
> 
> Also, could you please add tests with explicit `ansible-galaxy collection list -o human`? I believe I've only seen an implicit one.

This is done.

samdoran commented on lib/ansible/cli/galaxy.py:
```diff
@@ -301,6 +302,10 @@ def add_list_options(self, parser, parents=None):
         list_parser.add_argument(galaxy_type, help=galaxy_type.capitalize(), nargs='?', metavar=galaxy_type)
 
+        if galaxy_type == 'collection':
+            list_parser.add_argument('--format', dest='output_format', choices=('human', 'yaml', 'json'), default='human',
```
Do we need to specify `human` as an option? That could just the default and `yaml` or `json` are specified.

samdoran replied:
This is not a blocker, just a thought on the interface.

shanemcd replied:
@samdoran This was requested here: https://github.com/ansible/ansible/pull/73474#discussion_r569948199 I have also added a test for this functionality.

 I don't have a strong opinion either way. I will defer to the group. @s-hertel @webknjaz @jborean93 @bcoca - thoughts?

webknjaz replied:
I was unable to find examples of a `default` that is not listed in `choices` on the internet. This may need to be checked experimentally.

webknjaz replied:
@shanemcd my understanding is that @sdoran suggests
```suggestion
            list_parser.add_argument('--format', dest='output_format', choices=('yaml', 'json'), default='human',
```
and it's undocumented whether this would work argparse-wise.

shanemcd replied:
It _does_ seem to work. Again, I have no strong feelings here. I'm happy to make this change + delete the explicit 'human' format test if that's what folks want.

shanemcd replied:
Talked with Sam and he said this is not a blocker. Just going to leave in this way.

samdoran commented on lib/ansible/cli/galaxy.py:
```diff
@@ -1466,6 +1487,11 @@ def execute_list_collection(self, artifacts_manager=None):
         for w in warnings:
             display.warning(w)
 
+        if output_format == 'json':
+            display.display(json.dumps(collections_in_paths))
+        elif output_format == 'yaml':
+            display.display(yaml.safe_dump(collections_in_paths))
+
```
If no collection paths are found, this will output an empty `{}` before the help and error message. This should be after the `raise AnsibleOptionsError` below.

```
> ANSIBLE_COLLECTIONS_PATH='' ansible-galaxy collection list --format json
{}
usage: ansible-galaxy [-h] [--version] [-v] TYPE ...

Perform various Role and Collection related operations.

positional arguments:
  TYPE
    collection   Manage an Ansible Galaxy collection.
    role         Manage an Ansible Galaxy role.

optional arguments:
  --version      show program's version number, config file location, configured module search path, module location, executable location and exit
  -h, --help     show this help message and exit
  -v, --verbose  verbose mode (-vvv for more, -vvvv to enable connection debugging)
ERROR! - None of the provided paths were usable. Please specify a valid path with --collections-path
```

webknjaz replied:
Why? Doesn't it go to stdout while help goes to stderr? With typical piping, it'd be alright and the tools down the chain would be able to consume and empty "there's nothing installed" `{}`...

shanemcd replied:
Awesome find. Fixed + added test.

shanemcd replied:
The empty JSON object was being included in stdout along with the usage text. IMO this _was_ a bug / undesirable behavior.

samdoran replied:
I didn't know if it was intentional to output an empty dict. I don't think it will cause issues, just seems like it shouldn't be doing that.

shanemcd replied:
Demonstration:

```
$ ANSIBLE_COLLECTIONS_PATH='' ansible-galaxy collection list --format yaml 2> /dev/null
{}
usage: ansible-galaxy [-h] [--version] [-v] TYPE ...

Perform various Role and Collection related operations.

positional arguments:
  TYPE
    collection   Manage an Ansible Galaxy collection.
    role         Manage an Ansible Galaxy role.

optional arguments:
  --version      show program's version number, config file location, configured module search path, module location, executable location and exit
  -h, --help     show this help message and exit
  -v, --verbose  verbose mode (-vvv for more, -vvvv to enable connection debugging)
```

shanemcd replied:
Some options I can identify:

- Leave it as-is (patch w/ Sam's suggestion)
- We stop using `AnsibleOptionsError` here: https://github.com/ansible/ansible/blob/649df97d9174a285bcb08b881c45d2af4058d450/lib/ansible/cli/galaxy.py#L1491
- Make `AnsibleOptionsError` send to stderr
- We change the conditional that throws the error to be `if not path_found and output_format != 'human'`.

samdoran replied:
The use of `AnsibleOptionsError` is to make in consistent with other sub commands. 

It looks like the help and `{}` will be it stdout. The error message is in `stderr`.

samdoran replied:
So the fact that the help is in the `stdout` could cause issues. If the help went to `stderr`, that would be better in this case, but I don't think we should change that.

shanemcd replied:
Ok. Let's roll with it how it is now.

shanemcd committed:
Fix bug where empty json object was emitted + add test

shanemcd commented:
Quick, someone push the button before I break the CI again!
